test/test_nn.py

```diff
@@ -2622,8 +2765,8 @@ def test_pruning_serialization_model(self):
         # check that the original weight and the new mask are present
         self.assertIn('0.weight_orig', model.state_dict())
         self.assertIn('0.weight_mask', model.state_dict())
-        self.assertNotIn('0.weight', model.state_dict())
-        self.assertTrue(hasattr(model[0], 'weight'))
+        # Weight is now a buffer
```

Comment:
True. How should we go about this? In my opinion, it makes more sense to have `weight` as a buffer, and having it serialized with the whole model, given that tensor hooks (I don't know about module hooks) are not serialized.

Reply:
The original (unpruned) weight and the mask are everything you need to recreate the pruned weight by simply multiplying the two together [true in pruning, not sure about other use cases]. So, in the pruning case, there would be no need for the weight to also be stored in the state_dict for saving -- it would just be redundant info. And this would be another copy of the whole net if all layers are at least partially pruned.

In general, though, wouldn't most users want to make the parametrization permanent (through `.remove`) before serializing?

Reply:
I agree that it is somehow redundant, but it is not entirely It is not entirely redundant when saving the module, since the tensor hooks, at the moment, are not serialized. As such, if the user wants to save and resume later the training, they will have to set the parametrization again. In this case, the saved buffer can be used as a check to see that the saving / loading has been performed correctly. My reasoning behind putting them as a buffer was that this way it is more orthogonal with the way pytorch works. For example, they are moved to GPU when the user chooses to move the whole model to GPU. Otherwise this would have to be done by hand.

I do not think that one would want to make parametrizations permanent before serializing in general. In the general case, a parametrization is a constraint on certain parameter. For this reason, one may want to save a checkpoint of a model and resume the training later. To be able to do so, one needs the parametrization and the original tensors to resume the optimization with the constraint.